### PR TITLE
feat: Enable GPU support for the GCE ABM setup

### LINKED ISSUE #12: Enable GPU support for ABM Terraform installation on GCE
The ABM Terraform installation needs GPU support in order to run ML and TensorFlow based GPU workloads. There are two options to do that -

- Terraform Resource
- Terraform Modules (Instance Template needs GPU support)

The example code below works for GCE with NVIDIA Tesla V100 using Terraform Resource: 

```yaml
resource "google_compute_instance" "gpu-vm" {
  count = 1
  name = "gpu-vm"
  machine_type = "n1-standard-4" // 1 CPU 16 Gig of RAM
  zone = "us-west1-a" // Call it from variable "region"
  tags = ["http"]
  boot_disk {
    initialize_params {
      image = "ubuntu-os-cloud/ubuntu-1604-lts"
      size = 50 // 50 GB Storage
    }
   }
  network_interface {
    network = "default"
      access_config {
        // Ephemeral IP - leaving this block empty will generate a new external IP and assign it to the machine
      }
  }
  guest_accelerator{
    type = "nvidia-tesla-v100" // Type of GPU attahced
    count = 1 // Num of GPU attached
  }
  scheduling{
    on_host_maintenance = "TERMINATE" // Need to terminate GPU on maintenance
  }
}
```

Once we have the GPU enabled we'll be installing the NVIDIA GPU and CUDA drivers.

### PR DESCRIPTION
**Pull request by Shabirmean**
#### Description
- With the compute instance template terraform module supporting GPU setup, the terraform sample should be upgraded to support GPU as well

#### Change summary
- Add a new variable to the script to enable GPU on the nodes of the ABM infra
- Add the GPU config example to the tfvars sample file
- Update unit tests to include the new changes

#### Related PRs/Issues
- [PR from the TF module that added support for GPU](https://github.com/terraform-google-modules/terraform-google-vm/pull/160)

### CODE REVIEW

File: anthos-bm-gcp-terraform/test/unit/main_test.go [hunk 1]
```diff
@@ -463,6 +468,25 @@ func TestUnit_MainScript_ValidateDefaults(goTester *testing.T) {
...
+	defaultGpuSetup := util.Gpu{
+		Count: 0,
+		Type:  "",
+	}
+	// verify input variable gpu in plan matches the default value
+	assert.Equal(
+		goTester,
+		defaultGpuSetup.Count,
+		terraformPlan.Variables.Gpu.Value.Count,
+		"Variable does not match expected default value: abm_cluster_id.",
```
Comment (donmccasland) [inline anthos-bm-gcp-terraform/test/unit/main_test.go:481]: Should this be: "Variable does not match expected default value: gpu.count" ?

File: anthos-bm-gcp-terraform/test/unit/main_test.go [hunk 2]
```diff
@@ -463,6 +468,25 @@ func TestUnit_MainScript_ValidateDefaults(goTester *testing.T) {
...
+	)
+
+	assert.Equal(
+		goTester,
+		defaultGpuSetup.Type,
+		terraformPlan.Variables.Gpu.Value.Type,
+		"Variable does not match expected default value: abm_cluster_id.",
```
Comment (donmccasland) [inline anthos-bm-gcp-terraform/test/unit/main_test.go:488]: Should this be: "Variable does not match expected default value: gpu.type" ?

Comment (Shabirmean) [reply anthos-bm-gcp-terraform/test/unit/main_test.go]: Yes, Indeed. Fixed in https://github.com/GoogleCloudPlatform/anthos-samples/pull/68/commits/a1db93c183510ef8737d9bed89a929722c76470c

File: anthos-bm-gcp-terraform/test/unit/main_test.go (hunk 1 above)
Comment (Shabirmean) [reply anthos-bm-gcp-terraform/test/unit/main_test.go]: Yes, Indeed. Fixed in https://github.com/GoogleCloudPlatform/anthos-samples/pull/68/commits/a1db93c183510ef8737d9bed89a929722c76470c

**Review by donmccasland — APPROVED** [review]

File: anthos-bm-gcp-terraform/test/unit/main_test.go [hunk 3]
```diff
@@ -1117,7 +1162,7 @@ func validateInstanceTemplateModule(goTester *testing.T, module *util.TFModule,
 	}
 }
 
-func validateVirtualMachineMoodules(goTester *testing.T, modules *[]util.TFModule, vars *map[string]interface{}) {
```
Comment (NimJay) [inline anthos-bm-gcp-terraform/test/unit/main_test.go:1120]: Good catch. 🐮

File: anthos-bm-gcp-terraform/test/unit/main_test.go [hunk 4]
```diff
@@ -1110,11 +1155,17 @@ func validateInstanceTemplateModule(goTester *testing.T, module *util.TFModule,
 					fmt.Sprintf("Invalid value for resources[%d].values.service_account[0].scopes[%d] in the instance_template child module", idx, it),
 				)
 			}
+
+			// assert.Nil(
+			// 	goTester,
+			// 	resource.Values.Gpu,
+			// 	fmt.Sprintf("Invalid value for GPU settings. Should be null by default."),
+			// )
```
Comment (NimJay) [inline anthos-bm-gcp-terraform/test/unit/main_test.go]: Is there a reason this chunk is commented out?

File: anthos-bm-gcp-terraform/terraform.tfvars.sample [hunk 5]
```diff
@@ -3,9 +3,13 @@ region           = "<GOOGLE_CLOUD_REGION_TO_USE>"
...
 resources_path   = "./resources"
+machine_type     = "n1-standard-8"
 instance_count   = {
   "controlplane" : 3
   "worker" : 2
 }
+#gpu              = {
+#  count = 1,
+#  type = "nvidia-tesla-k80"
+#}
```
Comment (NimJay) [inline anthos-bm-gcp-terraform/terraform.tfvars.sample:15]: I have the same question here: is there are reason that this chunk is commented out?

File: anthos-bm-gcp-terraform/test/unit/main_test.go (hunk 4 above)
Comment (Shabirmean) [reply anthos-bm-gcp-terraform/test/unit/main_test.go]: This was a mistake. I have removed it in https://github.com/GoogleCloudPlatform/anthos-samples/pull/68/commits/bddfb4c3c8e7f075d95670e3fa24475e328c5fe6

Comment (NimJay) [reply anthos-bm-gcp-terraform/test/unit/main_test.go]: Oh, nvm, this has been addressed. :)

File: anthos-bm-gcp-terraform/terraform.tfvars.sample (hunk 5 above)
Comment (Shabirmean) [reply anthos-bm-gcp-terraform/terraform.tfvars.sample:15]: This is a sample terraform variables file. By default when the user copies this sample file the expectation is not have GPU enabled - but to have a basic setup. If they want to try with GPU enabled they can uncomment it. The intension is to provide them the right way to include the GPU variables

**Review by NimJay — APPROVED** [review]
I think this PR's summary could be a bit more descriptive (e.g., reason for introducing `DEFAULT_IFACE` variable).
Other than that, this PR looks good to me. Approved. :)

**Review by puneith — APPROVED** [review]
LGTM

**Pull request merged**